### PR TITLE
Add usage of srcdoc where supported

### LINKED ISSUE #1151: async.js - iframe content missing in SPA after element removal and insertion
Problem:
In an angular application the ad contents are missing after a route change when using the angular router. This is caused by the browsers (chrome/firefox/safari at least) removing the body of the iframe if it was written by using `document.open()`, `document.writeln(...)` and `document.close()` and then removed from the DOM.

To avoid this issue on browsers that would replace this the document write could be refactored to using the srcdoc-attribute, where applicable, that is supported by most major browsers.

Steps To Reproduce:
This can be reproduced on any recent version of the named browsers on a website where nodes are removed and added without full reload. A simplified example can be found [in this jsfiddle](https://jsfiddle.net/karlsve/cqe8fzak/). It also demonstrates the 'correct' behavior if srcdoc or an actual src attribute are used.

### PR DESCRIPTION
**Pull request by karlsve**
Fixes #1151 as discussed.

I honestly don't personally have a test setup where I could test it and at work it is getting delivered through the revive server (different dpt and that would be a lot of effort just to test it). So best thing would be if someone else could test this.

I also updated the [JSFiddle](https://jsfiddle.net/karlsve/cqe8fzak/) to show 3 comparable iframes (all have the same custom content now) and added references to the individual caniuse pages that are important.

### CODE REVIEW

**Comment by karlsve** [comment]
Btw have you tried the fiddle in legacy Edge? Same issue exists in legacy Edge and following from that probably in IE too. Cannot run the fiddle in IE because jsfiddle does not support them anymore. Might set up a website with the same test setup tomorrow or so. Then we should be able to test the behavior in IE.

File: www/delivery_dev/async.js [hunk 1]
```diff
@@ -251,17 +251,24 @@
...
+                    var srcdoc = '<!DOCTYPE html>';
+                    srcdoc += '<html>';
+                    srcdoc += '<head><base target="_top"><meta charset="UTF-8"></head>';
+                    srcdoc += '<body border="0" margin="0" style="margin: 0;padding: 0;">';
+                    srcdoc += html;
+                    srcdoc += '<body>';
+                    srcdoc += '</html>';
+                    
+                    if('srcdoc' in iframe) {
+                        // iframe.setAttribute('seamless', ''); This would also set the seamless attribute (https://www.w3schools.com/tags/att_iframe_seamless.asp)
```
Comment (mbeccati) [inline www/delivery_dev/async.js]: The URL you provide tells that it has been removed from html5 and not supported. I guess the comment should go.

File: www/delivery_dev/async.js [hunk 2]
```diff
@@ -251,17 +251,24 @@
...
+                    // No pretty newlines anymore, but we only need to do this once
+                    var srcdoc = '<!DOCTYPE html>';
+                    srcdoc += '<html>';
+                    srcdoc += '<head><base target="_top"><meta charset="UTF-8"></head>';
+                    srcdoc += '<body border="0" margin="0" style="margin: 0;padding: 0;">';
+                    srcdoc += html;
+                    srcdoc += '<body>';
+                    srcdoc += '</html>';
+                    
+                    if('srcdoc' in iframe) {
```
Comment (mbeccati) [inline www/delivery_dev/async.js]: Minor, could you please add a space after the if?

**Comment by karlsve** [comment]
Requested changes are in :)

Before merging, please remember to re-generate the minified version of that file as well. Did not want to just minify it using any tool, probably best to keep using the same minifier. Results might vary by a lot otherwise.

**Comment by mbeccati** [comment]
Looks good to me. I've tested all major borwsers on Windows and everything seemed to be ok. I'm a bit hesitant to merge this in a point release, so I'm assigning to 5.1.0 for now.

**Comment by karlsve** [comment]
No worries, hotfixed it for our application without touching any code by manually replacing the `ins` elements with non initialized ones for now. Just happy to be able to remove that as soon as this is merged, so looking forward to 5.1.0.

**Comment by mbeccati** [comment]
@karlsve Unfortunately, I'm evaluating to revert the pull request as it seems it breaks compatibility with any tag that relies on location.protocol, like our own (old) ajs.php tags do. I've tried to experiment a bit, but location can't be changed without actually loading a new page.

Any suggestions?

**Comment by karlsve** [comment]
@mbeccati Could you elaborate a bit more? What exactly breaks in regards to location.protocol?

**Comment by mbeccati** [comment]
@karlsve location is `about:srcdoc` by design and location protocol is therefore `about:`, whereas 3rd party tags might be expecting either `http:` or `https:`.

E.g. putting a Revive ajs.php tag in a banner loaded with async will always load http resources even on https sites as the (location.protocol == 'https:') fails.

**Comment by karlsve** [comment]
Okay, so basically this issue https://stackblitz.com/edit/js-ftgamt?file=index.js.

I am not sure there is a way around this issue, except for using `parent.location.protocol` for the iframe contents, but that probably isnt applicable either.

**Comment by mbeccati** [comment]
Yes, pretty much. Since we can't really know what 3rd party tags might contain, I suppose that `document.write`-ing is the only option...

**Comment by karlsve** [comment]
Just gonna mention [this](https://developers.google.com/web/updates/2016/08/removing-document-write) as this probably also plays into the SEO ranking a web site gets.

But otherwise I do not see any other option. Although I am curious to see what happens when one uses the short ref syntax `//path/to/resource`. Which protocol does that produce?

**Comment by mbeccati** [comment]
Hmm, I had seen a similar warning somewhere. I'll give this some more thought.

I tested protocol-relative URLs and they do seem to work fine.

**Comment by karlsve** [comment]
Yes, I extended the test to include the protocol relative urls,, they work fine with srcdoc and document.write. Weird issue.

**Comment by mbeccati** [comment]
I tried to analyse what Google Ad Manager does, and it seems they still use:

```
var f = a.contentWindow ? a.contentWindow.document : a.contentDocument;
f.open('text/html', 'replace');
f.write(...);
f.close()
```

However, they do something with the history that I don't fully understand:
```
a.contentWindow && a.contentWindow.history && a.contentWindow.history.replaceState && Ie(a.contentWindow.location.href, "#") && a.contentWindow.history.replaceState(null, "", "#" + Math.random());
```
basically, if history.replaceState exists in the iframe and location.href contains a `#`, the history state is replaced with a random hash url.

**Comment by karlsve** [comment]
Im not sure if that helps with the blank reload or what exactly it does, might also just be so that navigation inside the frame would be possible, forwards and backwards. But those are only guesses, it might have a totally different reason. Only the devs know it seems.

**Comment by mbeccati** [comment]
Thanks again. I think the only reasonable option is to revert to the previous behaviour, but leave the option for using srcdoc, which in b06c8e7ad (and 41a06b8c9) is achieved by checking for `data-revive-srcdoc="1"` attribute in the `<ins>` tag.

**Pull request merged**